src/renderer/dx/CustomTextLayout.cpp

```diff
@@ -170,21 +167,30 @@ CATCH_RETURN()
         UINT32 uiLengthRead = 0;
 
         // Start from the beginning.
-        const UINT32 glyphStart = 0;
+        UINT32 pos = 0;
 
         _glyphIndices.resize(textLength);
 
-        const HRESULT hr = _fontRenderData->Analyzer()->GetTextComplexity(
-            _text.c_str(),
-            textLength,
-            _fontInUse,
-            &isTextSimple,
-            &uiLengthRead,
-            &_glyphIndices.at(glyphStart));
-
-        RETURN_IF_FAILED(hr);
-
-        _isEntireTextSimple = isTextSimple && uiLengthRead == textLength;
+        while (pos < textLength)
+        {
+            const HRESULT hr = _fontRenderData->Analyzer()->GetTextComplexity(
+                &_text.at(pos),
+                textLength,
+                _fontInUse,
+                &isTextSimple,
+                &uiLengthRead,
+                &_glyphIndices.at(pos));
+
+            RETURN_IF_FAILED(hr);
+            _SetCurrentRun(pos);
+            _SplitCurrentRun(pos);
+            pos += std::max(uiLengthRead, 1u);
+            while (uiLengthRead > 0)
+            {
+                auto& run = _FetchNextRun(uiLengthRead);
+                run.isTextSimple = isTextSimple;
+            }

```

Comment:
nit: This might deserve a comment calling out that _FetchNextRun decrements uiLengthRead. I thought this was an infinite loop!